### PR TITLE
Introduce transaction outcome data type for recipts

### PR DESCRIPTION
**Pull request by agnikowa**
should be merged after: #511 and #512

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala [hunk 1]
```diff
@@ -167,10 +167,15 @@ object PV63 {
...
-    implicit class ReceiptEnc(msg: Receipt) extends RLPSerializable {
+    implicit class ReceiptEnc(receipt: Receipt) extends RLPSerializable {
       override def toRLPEncodable: RLPEncodeable = {
-        import msg._
-        RLPList(postTransactionStateHash, cumulativeGasUsed, logsBloomFilter, RLPList(logs.map(_.toRLPEncodable): _*))
+        import receipt._
+        val stateHash = postTransactionStateHash match {
+          case HashOutcome(hash) => hash.toArray[Byte]
+          case SuccessOutcome => Array(1.toByte)
+          case _ => Array(0.toByte)
```
Comment (mmrozek) [inline src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala]: Shouldn't it be an empty array?

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala [hunk 2]
```diff
@@ -190,7 +195,12 @@ object PV63 {
     implicit class ReceiptRLPEncodableDec(val rlpEncodeable: RLPEncodeable) extends AnyVal {
       def toReceipt: Receipt = rlpEncodeable match {
         case RLPList(postTransactionStateHash, cumulativeGasUsed, logsBloomFilter, logs: RLPList) =>
-          Receipt(postTransactionStateHash, cumulativeGasUsed, logsBloomFilter, logs.items.map(_.toTxLogEntry))
+          val stateHash = postTransactionStateHash match {
+            case RLPValue(bytes) if bytes.length > 1 => HashOutcome(ByteString(bytes))
+            case RLPValue(bytes) if bytes.length == 1 && bytes.head == 1 => SuccessOutcome
+            case _ => FailureOutcome
```
Comment (mmrozek) [inline src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala]: I prefer to write explicit all meaningful cases. In this situation value different than RLPValue is treated as a FailureOutcome.

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala (hunk 1 above)
Comment (agnikowa) [reply src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala]: you're right, changed.

**Comment by agnikowa** [comment]
there are 3 tests for it in `LedgerSpec` class https://github.com/input-output-hk/mantis/pull/513/files#diff-e44b517ae530cddefaa864eb1c98cda5R690 or do you mean something more?

**Comment by mmrozek** [comment]
Sorry, I missed that. It should be enough.

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala [hunk 3]
```diff
@@ -167,10 +167,15 @@ object PV63 {
   object ReceiptImplicits {
     import TxLogEntryImplicits._
 
-    implicit class ReceiptEnc(msg: Receipt) extends RLPSerializable {
+    implicit class ReceiptEnc(receipt: Receipt) extends RLPSerializable {
       override def toRLPEncodable: RLPEncodeable = {
-        import msg._
-        RLPList(postTransactionStateHash, cumulativeGasUsed, logsBloomFilter, RLPList(logs.map(_.toRLPEncodable): _*))
+        import receipt._
+        val stateHash = postTransactionStateHash match {
```
Comment (KonradStaniec) [inline src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala]: here you can add type to stateHash like : `val stateHash: RLPEncodeable`, and then in cases you do not need to serialize it by hand i.e you can just write:

```
case HashOutcome(hash) => hash
case SuccessOutcome => 1.toByte      
case _ => 0.toByte
```
And proper serialization implicitis will kick in.

**Pull request merged**